### PR TITLE
Convert the DES weak key and key parity tests to be constant time.     

### LINKED ISSUE #16859: Some uses of DES miscompute with OpenSSL 3.0.0
I am concerned that OpenSSL's deprecation of DES appears to cause silent miscomputation, which e.g. in JtR we only detect due to our runtime tests against our included test vectors:

https://github.com/openwall/john/issues/4831#issuecomment-945964561

Wouldn't in some other application this result in encryption silently becoming a no-op, which is generally a lot worse than 3DES?

### PR DESCRIPTION
**Pull request by paulidale**
Fixes #16944
Fixes #16859

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/des/set_key.c [hunk 1]
```diff
@@ -62,15 +64,18 @@ void DES_set_odd_parity(DES_cblock *key)
...
     unsigned int i;
+    unsigned char res = 0377;
 
-    for (i = 0; i < DES_KEY_SZ; i++) {
-        if ((*key)[i] != odd_parity[(*key)[i]])
-            return 0;
-    }
-    return 1;
+    for (i = 0; i < DES_KEY_SZ; i++)
+        res &= constant_time_eq_8((*key)[i], odd_parity[(*key)[i]]);
```
Comment (bernd-edlinger) [inline crypto/des/set_key.c:77]: accessing the odd_parity array with a secret key byte is actually not constant time

Comment (bernd-edlinger) [reply crypto/des/set_key.c:77]: its probably overkill but you could do something like that in constant time:
```C
  unsigned b = (*key)[i];
  b ^= b >> 4;
  b ^= b >> 2;
  b ^= b >> 1;
  res &= constant_time_eq(b & 1, 1);
```

Comment (mattcaswell) [reply crypto/des/set_key.c:77]: Personally I prefer this over the table look up approach

Comment (paulidale) [reply crypto/des/set_key.c:77]: Okay, I'll look at changing this tomorrow.  @bernd-edlinger, I was aware that the parity array look up wouldn't be constant time.  I expected it to be far less gratuitous than what was present.

BTW, gcc and presumably other compilers have a `__builtin_parity` intrinsic.

When are compilers going to add a _constant_time_ attribute.  I expect the [movcc](https://github.com/xoreaxeaxeax/movfuscator) would come close for all inputs.

Comment (paulidale) [reply crypto/des/set_key.c]: Updated to use the above code instead of a table lookup.

Comment (paulidale) [reply crypto/des/set_key.c]: I think it should be possible to compute all of the parities in parallel.  I doubt it's worth the effort for DES though.

**Review by bernd-edlinger — APPROVED** [review]

**Comment by solardiz** [comment]
Good stuff. However, while this fixes the internals of `DES_check_key_parity` and `DES_is_weak_key`, it does not yet fix the two `if` statements in `DES_set_key` (potential branching, unless compiled into `cmov` instructions or such). I think ideally those would be replaced with arithmetic/bitwise operations to compute the return value.

**Comment by paulidale** [comment]
Merged to master and 3.0.  Thanks for the reviews and suggestions.

As for the timing attack against the out if statement, I'm unconvinced.  An attacker would know you're using one of the know weak keys or that there is a parity error somewhere.  The former is a complete compromise but it's easy to check for anyway, the latter doesn't seem to reveal anything useful.

**Comment by solardiz** [comment]
Not a big deal, but FWIW:

> that there is a parity error somewhere [...] doesn't seem to reveal anything useful.

It might if the extra bits treated as parity are correlated with the actual 56 bits of key, such as when a human-chosen password is used as key.

I've just checked that inside OpenSSL itself all of `DES_fcrypt` and `DES_string_to_key` and `DES_string_to_2keys` use `DES_set_key_unchecked`. So we're fine there. However, some external implementation of these or similar (e.g., in an implementation of Kerberos or VNC) might be affected.

**Pull request closed without merging**